File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    List<TreeNode> pathP;
    List<TreeNode> pathQ;
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if(root == null || root == p || root == q){
            return root;
        }

        this.pathP = new ArrayList<>();
        this.pathQ = new ArrayList<>();

        backtrack(root, p, q, new ArrayList<>());
        //we can iterate on both paths

        for(int i = 0; i < pathP.size(); i++){
            if(pathP.get(i).val != pathQ.get(i).val){
                return pathP.get(i - 1);
            }
        }
        return null;
    }

    private void backtrack(TreeNode root, TreeNode p, TreeNode q, List<TreeNode> path){
        //base case
        if(root == null){
            return;
        }
        //logic
        if(root.val == p.val){
            path.add(p);
            path.add(p);
            pathP = new ArrayList<>(path);
            path.remove(path.size() - 1);
            path.remove(path.size() - 1);
        }

        if(root.val == q.val){
            path.add(q);
            path.add(q);
            pathQ = new ArrayList<>(path);
            path.remove(path.size() - 1);
            path.remove(path.size() - 1);
        }

        //action 
        path.add(root);
        backtrack(root.left, p, q, path);
        backtrack(root.right, p, q, path);
        //backtrack
        path.remove(path.size() - 1);
    }
}

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    pathP = []
    pathQ = []
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        if root is None or p.val is None or q.val is None:
            return root
        
        self.pathP = []
        self.pathQ = []

        self.dfs(root, p, q, path= list())
        # print(self.pathP)
        # print(self.pathQ)
        for i in range(len(self.pathP)):
            if self.pathP[i].val != self.pathQ[i].val:
                return self.pathP[i -1]
        
        return None
    
    def dfs(self, root, p, q, path):
        #base case
        if root is None:
            return

        if root.val == p.val:
            # print("this")
            self.pathP = path.copy()
            self.pathP.append(root)
            self.pathP.append(root)
            # print(self.pathP)


        elif root.val == q.val:
            self.pathQ = path.copy()
            self.pathQ.append(root)
            self.pathQ.append(root)

        #logic
        # print(path)
        #recurse
        path.append(root)
        self.dfs(root.left, p, q, path)
        #backtrack
        
        self.dfs(root.right, p, q, path)
        path.pop(-1)

